### PR TITLE
z3: Add output for Java bindings (*.jar and *.so)

### PR DESCRIPTION
**Pull request by lorenzleutgeb**
###### Motivation for this change

I couldn't find `com.microsoft.z3.jar` and `libz3java.so`, then I found out that they ended up in the `lib` output. I think having a separate output is better.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by lorenzleutgeb** [comment]
Ping @ttuegel

File: pkgs/applications/science/logic/z3/default.nix [hunk 1]
```diff
@@ -49,9 +49,14 @@ stdenv.mkDerivation rec {
     mkdir -p $python/lib
     mv $lib/lib/python* $python/lib/
     ln -sf $lib/lib/libz3${stdenv.hostPlatform.extensions.sharedLibrary} $python/${python.sitePackages}/z3/lib/libz3${stdenv.hostPlatform.extensions.sharedLibrary}
+  '' + optionalString javaBindings ''
+    mkdir -p $java
+    mv com.microsoft.z3.jar $java
+    mv $lib/lib/libz3java.${stdenv.hostPlatform.extensions.sharedLibrary} $java
```
Comment (ttuegel) [inline pkgs/applications/science/logic/z3/default.nix:55]: This is the preferred way to move files to another output:

```suggestion
    moveToOutput com.microsoft.z3.jar $java
    moveToOutput lib/libz3java.${stdenv.hostPlatform.extensions.sharedLibrary} $java
```

Are you sure the JAR file should end up in the root of the output?

**Review by ttuegel — CHANGES_REQUESTED** [review]

File: pkgs/applications/science/logic/z3/default.nix (hunk 1 above)
Comment (lorenzleutgeb) [reply pkgs/applications/science/logic/z3/default.nix]: Thanks for introducing me to `moveToOutput`.

I did some more searching, and found [`addPkgToClassPath`](https://github.com/NixOS/nixpkgs/blob/e7fe577d9803885d1191c6612b95c246cb605dde/pkgs/build-support/setup-hooks/set-java-classpath.sh#L6-L11) which suggests to use `$java/share/java/com.microsoft.z3.jar`.

Fixed in 6f4fe079422.

**Review by ttuegel — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Please do not merge master into PRs to resolve merge conflicts.

**Comment by ttuegel** [comment]
> Please do not merge master into PRs to resolve merge conflicts.

@SuperSandro2000 Where may I read about this policy?

**Comment by SuperSandro2000** [comment]
> > Please do not merge master into PRs to resolve merge conflicts.
> 
> @SuperSandro2000 Where may I read about this policy?

This is a general anti pattern not specific to nix because it usually obfuscates the git history and makes it hard to follow changes back because the history is no longer linear.

**Comment by ttuegel** [comment]
> This is a general anti pattern not specific to nix because it usually obfuscates the git history and makes it hard to follow changes back because the history is no longer linear.

@SuperSandro2000 Can you please show me where the Nixpkgs project has adopted this policy?

If this policy has been adopted by the organization, then an administrator can disable the feature in the repository settings. This will prevent confusion with new contributors, and save you a lot of work besides.

**Comment by SuperSandro2000** [comment]
> > This is a general anti pattern not specific to nix because it usually obfuscates the git history and makes it hard to follow changes back because the history is no longer linear.
> 
> @SuperSandro2000 Can you please show me where the Nixpkgs project has adopted this policy?
> 
> If this policy has been adopted by the organization, then an administrator can disable the feature in the repository settings. This will prevent confusion with new contributors, and save you a lot of work besides.

This is not written anywhere down but just generally good manner for creating pull request on GitHub and a standard rule when working with git.
